### PR TITLE
[unitaryHACK] Added `__repr__` method for QNode

### PR DESCRIPTION
**Pull request by paniash**
**Context:**

Adds `__repr__` method to print useful information when running `print(QNode)`.

**Related GitHub Issues:** #1352

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @paniash for this contribution 🙂 Let us know when it’s ready for review, or alternatively if you have any questions.

By the way, don’t forget to update the [changelog](.github/CHANGELOG.md), with:

- A small summary of the change
- A link back to this PR
- Your name added to the contributors section

**Comment by paniash** [comment]
@josh146 I think it's ready for review. Let me know if any further changes are required. :-)

File: pennylane/qnode.py [hunk 1]
```diff
@@ -28,7 +28,12 @@
 from pennylane.operation import State
 
 from pennylane.interfaces.autograd import AutogradInterface, np as anp
-from pennylane.tape import JacobianTape, QubitParamShiftTape, CVParamShiftTape, ReversibleTape
+from pennylane.tape import (
+    JacobianTape,
+    QubitParamShiftTape,
+    CVParamShiftTape,
+    ReversibleTape,
+)
```
Comment (josh146) [inline pennylane/qnode.py:36]: @paniash since this line hasn't been modified, it looks like `black` might have been run with a different line setting 🤔 

Could you re-run `black`, but with the line length set to 100? This is the standard we use in the repository to minimize formatting changes:

```python
black -l 100 pennylane
```

File: pennylane/qnode.py [hunk 2]
```diff
@@ -182,6 +187,20 @@ def __init__(
...
+
+    def __repr__(self):
+        """String representation."""
+        detail = "<QNode: wires={}, device='{}', interface='{}', diff_method='{}'>"
+        return detail.format(
+            self.device.num_wires,
+            self.device.short_name,
+            self.interface,
+            self.diff_method,
+        )
```
Comment (josh146) [inline pennylane/qnode.py:202]: It looks like `__str__` and `__repr__` currently return the same thing?

If so, best to simply define `__repr__` and remove `__str__` for now :)

**Review by josh146 — CHANGES_REQUESTED** [review]

File: pennylane/qnode.py (hunk 2 above)
Comment (paniash) [reply pennylane/qnode.py:202]: Hi! Would you recommend adding some other information for `__str__`. At this point, I don't know what could be added to it to make it different than `__repr__`.

Comment (josh146) [reply pennylane/qnode.py]: @paniash it's a good question, I'm also unsure what extra information would go in `__str__` 🤔 This potentially needs to be considered more, so I suggest removing `__str__` for now.

**Comment by paniash** [comment]
@josh146 Could you help me with the failing tests? Not sure why they're failing.

**Comment by albi3ro** [comment]
Hi @paniash ! thanks for this PR.

It looks like your checks are failing because we have decreased coverage by adding in lines that are not getting tested.

The QNode tests are at [`tests/tape/test_qnode.py`](https://github.com/PennyLaneAI/pennylane/blob/master/tests/tape/test_qnode.py)

Could you add some tests for the `__repr__` function there?  You could either stick them at the bottom or wherever makes the most sense. You just need to create some simple qnode's and then compare its `__repr__` output to a fixed expected value.

File: pennylane/qnode.py (hunk 1 above)
Comment (josh146) [reply pennylane/qnode.py:36]: Hi @paniash, I'm unresolving this one, since it seems this change is still here 🤔 

Did running `black` with a line length of 100 not fix it?

File: tests/tape/test_qnode.py [hunk 3]
```diff
@@ -456,8 +471,8 @@ def circuit(p1, p2=y, **kwargs):
 
         result = draw(circuit)(p1=x, p3=z)
         expected = """\
- 0: ──RX(0.1)───RX(0.4)──╭C──╭┤ ⟨Z ⊗ X⟩ 
- 1: ──RY(0.06)───────────╰X──╰┤ ⟨Z ⊗ X⟩ 
+ 0: ──RX(0.1)───RX(0.4)──╭C──╭┤ ⟨Z ⊗ X⟩
+ 1: ──RY(0.06)───────────╰X──╰┤ ⟨Z ⊗ X⟩
```
Comment (josh146) [inline tests/tape/test_qnode.py:475]: Hi @paniash, could you revert these changes here that are unrelated to the test you added above? Thanks!

File: pennylane/qnode.py (hunk 1 above)
Comment (paniash) [reply pennylane/qnode.py:36]: @josh146 I ran `black -l 100 pennylane/qnode.py` and it shows "1 file left unchanged".

File: tests/tape/test_qnode.py (hunk 3 above)
Comment (paniash) [reply tests/tape/test_qnode.py:475]: Apologies for such formatting changes. By default, my editor removes any trailing whitespaces.

File: tests/tape/test_qnode.py [hunk 4]
```diff
@@ -259,6 +259,21 @@ def test_validate_adjoint_invalid_device(self):
...
+        dev = qml.device("default.qubit", wires=1)
+
+        def func(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        qn = qml.QNode(func, dev)
+
+        assert (
+            qn
```
Comment (albi3ro) [inline tests/tape/test_qnode.py:273]: `qn.__repr__()`? Does that work?

File: pennylane/qnode.py (hunk 1 above)
Comment (josh146) [reply pennylane/qnode.py:36]: Ah, that's very odd 🤔 
Could you try updating your version of black (`pip install black --upgrade`) and trying again?

Comment (josh146) [reply pennylane/qnode.py:36]: Oh, nevermind, the CI check below seems to be passing, so this should be fine 🙂

File: tests/tape/test_qnode.py (hunk 4 above)
Comment (josh146) [reply tests/tape/test_qnode.py:273]:
```suggestion
            qn.__repr__()
```

Comment (paniash) [reply tests/tape/test_qnode.py]: Sorry about that. It's fixed now.

**Comment by paniash** [comment]
It looks like the test is failing for the circuit diagrams.  Not sure what's the issue though. My editor merely removed the trailing whitespaces (is that the reason for failing tests?).

**Comment by josh146** [comment]
>My editor merely removed the trailing whitespaces (is that the reason for failing tests?).

Hi @paniash, unfortunately yes it is! The circuit diagrams contain trailing whitespace that is required for the equality assertion to pass. If you revert these changes, the tests should pass again 🙂

**Comment by paniash** [comment]
@josh146 Thanks for letting me know! I'll fix it asap. :-)

**Comment by paniash** [comment]
Realizing for the first time that whitespaces are very important 😃

File: .github/CHANGELOG.md [hunk 5]
```diff
@@ -62,10 +62,13 @@ X_test = np.random.random((5, 6))
 K_test = qml.kernels.kernel_matrix(X_train, X_test, kernel)
 ```
 
+- Added `__str__` and `__repr__` methods to display readable info when printing QNodes.
```
Comment (josh146) [inline .github/CHANGELOG.md:65]:
```suggestion
* QNodes now display readable information when in interactive environments or when printed.
```

**Review by josh146 — APPROVED** [review]
This looks good to go from my end now @paniash! 🎉

**Comment by paniash** [comment]
@josh146 @albi3ro Thanks a lot for the guidance and constructive comments. Really appreciate it 😃
